test/libsolidity/SyntaxTestHelper.cpp

```diff
@@ -0,0 +1,159 @@
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <boost/test/unit_test.hpp>
+#include <boost/filesystem.hpp>
+#include <test/TestHelper.h>
+#include <test/libsolidity/AnalysisFramework.h>
+#include <test/libsolidity/SyntaxTestHelper.h>
+#include <test/libsolidity/SyntaxTestParser.h>
+
+namespace dev
+{
+namespace solidity
+{
+namespace test
+{
+
+class SyntaxTester : public AnalysisFramework
+{
+public:
+	void runSyntaxTest(SyntaxTest const& _test)
+	{
+		std::vector<std::string> unexpectedErrors;
+		auto expectations = _test.expectations;
+		auto errorList = parseAnalyseAndReturnError(_test.source, true, true, true).second;
+
+		// checks for presence of errors and warnings with multiplicities
+		// could be modified to enforce the precise order as well
+		for (auto const& error: errorList)
+		{
+			std::string msg = errorMessage(*error);
+			bool found = false;
+			for (auto it = expectations.begin(); it != expectations.end(); ++it)
+				if (msg.find(it->second) != std::string::npos && error->typeName() == it->first)
+				{
+					found = true;
+					expectations.erase(it);
+					break;
+				}
+			if (!found)
+				unexpectedErrors.emplace_back(error->typeName() + ": " + msg);
+		}
+
+		if (!unexpectedErrors.empty())
+		{
+			std::string msg = "Unexpected error(s):\n";
+			for (auto const& error: unexpectedErrors)
+				msg += error + "\n";
+			BOOST_ERROR(msg);
+		}
+
+		if (!expectations.empty())
+		{
+			std::string msg = "Expected error(s) not present:\n";
+			for (auto const& expectation: expectations)
+				msg += expectation.first + ": " + expectation.second + "\n";
+			BOOST_ERROR(msg);
+		}
+	}
+private:
+	static std::string errorMessage(Error const& _e)
+	{
+		return _e.comment() ? *_e.comment() : "NONE";
+	}
+
+};
+
+int registerSyntaxTests(
+	boost::unit_test::test_suite& _suite,
+	boost::filesystem::path const& _basepath,
+	boost::filesystem::path const& _path
+)
+{
+	static SyntaxTestParser syntaxTestParser;
+	static SyntaxTester syntaxTester;
+
+	int numTestsAdded = 0;
+	boost::filesystem::path fullpath = _basepath / _path;
+	if (boost::filesystem::is_directory(fullpath))
+	{
+		boost::unit_test::test_suite* sub_suite = BOOST_TEST_SUITE(_path.filename().string());
+		for (auto &entry : boost::filesystem::directory_iterator(fullpath))
+			numTestsAdded += registerSyntaxTests(*sub_suite, _basepath, _path / entry.path().filename());
+		_suite.add(sub_suite);
+	}
+	else
+	{
+		_suite.add(boost::unit_test::make_test_case(
+			[fullpath] { syntaxTester.runSyntaxTest(syntaxTestParser.parse(fullpath.string())); },
```

Comment:
Does this caputure both `syntaxTester` and `syntaxTestParser`? This looks a little weird to me, to be honest.

Reply:
Why do they need to be static anyway? In other boost tests, the full fixture is re-created for each test case, I think we should follow that pattern.

Reply:
Although I'll change this, as already discussed earlier, I read up on lambda expressions and capturing in such cases; from http://en.cppreference.com/w/cpp/language/lambda:
> A lambda expression can use a variable without capturing it if the variable
> - is a non-local variable or has **static** or thread local storage duration (in which case the variable cannot be captured), or 
> - ...

So using the function scope static variables inside the lambda without capturing them is in fact conformant :-).

Reply:
Actually I just also read up on boost test fixtures and realized that AnalysisFramework in SolidityNameAndTypeResolution.cpp is defined as a test suite fixture, which is in fact **not** recreated for every test, but there is only one instance for the whole test suite.

From http://www.boost.org/doc/libs/1_59_0/libs/test/doc/html/boost_test/tests_organization/fixtures/case.html#boost_test.tests_organization.fixtures.case.fixture_for_a_complete_subtree:
> In other words unless reset by another BOOST_FIXTURE_TEST_SUITE or BOOST_FIXTURE_TEST_CASE definition the same fixture is assigned to all test cases of a test suite, including ones that belong to the sub test suites. 

So following the existing pattern would in fact be to use static variables here, so given that I tend to leave it at that... any thoughts?

Reply:
[removed last comment since I was mistaken - the boost tests in fact call the constructor and destructor of the fixture for every test case, so I will change this to be consistent with the existing pattern]